# Protein Structure Prediction
# Improving the Understanding of the Vaginal Microbiota through De Novo Structural Protein Prediction and Annotation
Repository with all scripts and data generated for my internship at the Computational Biology department of Institut Pasteur of Paris. </p></p>

This research internship project was carried out in the Metagenomic Signatures group, led by Sean Kennedy, in the framework of the InSPIRe project (Innovative Strategies for Perinatal Infection Risk-reduction).
While there is some knowledge about the composition of the different core vaginal microbiomes, the understanding of the bacteria-host interactions within the vaginal microbiota is hampered by an incomplete functional annotation. </p>
The aim of my work was to apply ground-breaking protein structure prediction techniques to improve the annotation of dominant vaginal microbiota species in order to better understand their features and adaptations to and impact on the host vaginal microbiota. </p>
The goal was to develop a systematic pipeline that allows the functional annotation of predicted ORFs in bacterial genomes using the AlphaFold2 and Foldseek implementations on Institut Pasteur HPC cluster. </p>
This pipeline can be further used to study the coding content of other specie, and can guide the following <i>in vitro</i> studies for the experimental validation of the predicted protein, as this de novo functional annotation strongly reduces the number of targets to test.  
